docs: external KMS configuration (#23600)
hashicorp/nomad: website/content/docs/configuration/keyring/gcpckms.mdx

---
layout: docs
page_title: GCP Cloud KMS - Keyrings - Configuration
description: >-
  The GCP Cloud KMS keyring configures Nomad to use GCP Cloud KMS as the keyring
  wrapping mechanism.
---

# `gcpckms` keyring

The GCP Cloud KMS keyring configures Nomad to use GCP Cloud KMS to wrap its
keyring. This example shows configuring GCP Cloud KMS keyring through the Nomad

> configuration file by providing all the required values:

Reviewer: ```suggestion
configuration file by providing all the required values.
```

```hcl
keyring "gcpckms" {
  active = true
  name   = "example"

  credentials = "/etc/nomad.d/nomad-user-creds.json"
  project     = "nomad-project"
  region      = "global"
  key_ring    = "nomad-keyring"
  crypto_key  = "nomad-key"
}
```

## `gcpckms` parameters

These parameters apply to the `keyring` stanza in the Nomad configuration file:

> - `credentials` `(string: <required>)`: The path to the credentials JSON file to
>   use. May be also specified by the `GOOGLE_CREDENTIALS` or
>   `GOOGLE_APPLICATION_CREDENTIALS` environment variable or set automatically if
>   running under Google App Engine, Google Compute Engine or Google Kubernetes
>   Engine.

Reviewer: ```suggestion
- `credentials` `(string: <required>)`: The path to the credentials JSON file to
  use. Alternately specify via the `GOOGLE_CREDENTIALS` or
  `GOOGLE_APPLICATION_CREDENTIALS` environment variable or set automatically if
  running under Google App Engine, Google Compute Engine, or Google Kubernetes
  Engine.
```

>   Engine.

Author: Oops, caught me copy-and-pasting from Vault's docs.

> - `project` `(string: <required>)`: The GCP project ID to use. May also be
>   specified by the `GOOGLE_PROJECT` environment variable.
> 

Reviewer: ```suggestion
- `project` `(string: <required>)`: The GCP project ID to use. Alternately specify via the `GOOGLE_PROJECT` environment variable.

```

- `region` `(string: <required>)`: The GCP region/location where the key ring

>   lives. May also be specified by the `GOOGLE_REGION` environment variable.

Reviewer: ```suggestion
  lives. Alternately specify via the `GOOGLE_REGION` environment variable.
```

- `key_ring` `(string: <required>)`: The GCP CKMS key ring to use.

- `crypto_key` `(string: <required>)`: The GCP CKMS crypto key to use for
  encryption and decryption.

## Authentication &amp; permissions

> Authentication-related values must be provided, either as environment

Reviewer: ```suggestion
You must provide authentication-related values either as environment
```

variables or as configuration parameters.

GCP authentication values:

- `GOOGLE_CREDENTIALS` or `GOOGLE_APPLICATION_CREDENTIALS`
- `GOOGLE_PROJECT`
- `GOOGLE_REGION`

> Note: The client uses the official Google SDK and will use the specified
> credentials, environment credentials, or [application default credentials][] in
> that order, if the above GCP specific values are not provided.

Reviewer: ```suggestion
Note: The client uses the official Google SDK and the specified
credentials, environment credentials, or [application default credentials][] in
that order, if you do not provide the aforementioned GCP-specific values. 
```

The service account needs the following minimum permissions on the crypto key:

```text
cloudkms.cryptoKeyVersions.useToEncrypt
cloudkms.cryptoKeyVersions.useToDecrypt
cloudkms.cryptoKeys.get
```

These permissions can be described with the following role:

```text
roles/cloudkms.cryptoKeyEncrypterDecrypter
cloudkms.cryptoKeys.get
```

> `cloudkms.cryptoKeys.get` permission is used for retrieving metadata information
> of keys from CloudKMS within this engine initialization process.

Reviewer: ```suggestion
Use `cloudkms.cryptoKeys.get` permission for retrieving metadata information
of keys from CloudKMS within this engine initialization process.
```

## Key rotation

This keyring supports rotating keys defined in Google Cloud KMS
[doc][rotating_keys]. Both scheduled rotation and manual rotation is supported
for CKMS since the key information. Old keys version must not be disabled or

> deleted and are used to decrypt older data. Any new or updated data will be

Reviewer: ```suggestion
deleted and are used to decrypt older data. Any new or updated data is
```
